File: src/lib/time_diff.rb
# frozen_string_literal: true

require "time"

# One minute in seconds.
ONE_MIN = 60

# One hour (60 minutes) in seconds.
ONE_HOUR = 60 * ONE_MIN

# One day (24 hours) in seconds.
ONE_DAY = 24 * ONE_HOUR

# One week (7 days) in seconds.
ONE_WEEK = 7 * ONE_DAY

# One month (4 weeks) in seconds.
ONE_MONTH = 4 * ONE_WEEK

# One year (52 weeks) in seconds.
ONE_YEAR = 52 * ONE_WEEK

# Represents the difference between two times.
class TimeDiff
  # @param [Time] start_time
  # @param [Time] end_time
  def initialize(start_time, end_time)
    raise "Start time is invalid." unless start_time.is_a? Time
    raise "End time is invalid." unless end_time.is_a? Time
    raise "The end time must be greater than the start time." if end_time < start_time

    # @type [Time]
    @start_time = start_time

    # @type [Time]
    @end_time = end_time

    @units = {year: ONE_YEAR, month: ONE_MONTH, week: ONE_WEEK, day: ONE_DAY, hour: ONE_HOUR, minute: ONE_MIN}
  end

  # The time difference rounded to the nearest second.
  def value
    Float(@end_time - @start_time).round
  end

  # Returns the time difference as a hash, with each key representing a unit of time (year, month, etc.).
  def to_h
    diff_secs = value
    breakdown = Hash.new(0)

    @units.each do |unit, secs|
      while diff_secs >= secs
        diff_secs -= secs
        breakdown[unit] += 1
      end
    end

    breakdown[:second] = diff_secs if diff_secs.positive?
    breakdown
  end

  # Returns the time difference as a string, broken down by units of time (years, months, etc.).
  def to_s
    if value.zero?
      "No difference."
    else
      to_h.map { |u, c| unit_to_s(u, c) }.join(", ")
    end
  end

  private

  # Converts a time difference to a string.
  # @param [Symbol | String] unit
  # @param [Integer] count
  def unit_to_s(unit, count)
    "#{count} #{unit}#{"s" if count != 1}"
  end
end

File: src/lib/time-diff.js
import units from "./units.js";

//#region Types

/**
 * @typedef {Object} TimeUnits
 * @property {number} year
 * @property {number} month
 * @property {number} week
 * @property {number} day
 * @property {number} hour
 * @property {number} minute
 * @property {number} second
 */

//#endregion

//#region Classes

/** Represents the difference between two times. */
export class TimeDiff {
  /** @type {Date} */
  #endTime;

  /** @type {Date} */
  #startTime;

  /** @type {Map<string, number>} */
  #units = new Map([
    ["year", units.year],
    ["month", units.month],
    ["week", units.week],
    ["day", units.day],
    ["hour", units.hour],
    ["minute", units.minute],
  ]);

  //#region Public API

  /**
   * @param {Date} startTime The start of the date range.
   * @param {Date} endTime The end of the date range.
   */
  constructor(startTime, endTime) {
    this.#validateTimes(startTime, endTime);
    this.#startTime = startTime;
    this.#endTime = endTime;
  }

  toString() {
    if (this.seconds == 0) {
      return "No difference";
    }

    return Object.entries(this.#timeToUnits())
      .filter((entry) => entry[1] > 0)
      .map(this.#entryToString)
      .join(", ");
  }

  /** The time difference rounded to the nearest second. */
  get seconds() {
    const diffMs = this.#endTime - this.#startTime;
    return Math.round(diffMs / 1000);
  }

  //#endregion

  /**
   * Converts a `TimeUnits` value to a string.
   * @param {[keyof TimeUnits, number]} entry
   */
  #entryToString(entry) {
    const [unitName, value] = entry;
    let str = `${value.toLocaleString()} ${unitName}`;

    if (value != 1) {
      str += "s";
    }

    return str;
  }

  /**
   * Tests that a time is valid.
   * @param {Date} time The time to validate.
   */
  #isValidTime(time) {
    return time instanceof Date && !isNaN(time.valueOf());
  }

  /**
   * Returns the time difference as a dictionary, with each key representing
   * a unit of time (year, month, etc.).
   * @returns {TimeUnits}
   */
  #timeToUnits() {
    const units = {};
    let diffSeconds = this.seconds;

    this.#units.forEach((seconds, unit) => {
      units[unit] = 0;

      while (diffSeconds >= seconds) {
        diffSeconds -= seconds;
        units[unit] += 1;
      }
    });

    units.second = diffSeconds >= 0 ? diffSeconds : 0;
    return units;
  }

  /**
   * Validates that the start and end times form a valid date range.
   * @param {Date} startTime The start of the date range.
   * @param {Date} endTime The end of the date range.
   * @throws If either the `startTime` or `endTime` are invalid.
   */
  #validateTimes(startTime, endTime) {
    if (!this.#isValidTime(startTime)) {
      throw new Error("Start time is invalid.");
    }

    if (!this.#isValidTime(endTime)) {
      throw new Error("End time is invalid.");
    }

    if (endTime < startTime) {
      throw new Error("The end time must be greater than the start time.");
    }
  }
}

//#endregion
